Add Preloader::Association::LoaderQuery

jhawthorn opened the pull request:
Follow up to #41385 

This class represents the query a `Preloader::Association` uses to load its records. This replaces the previous concept of a grouping_key and serves both to group `Preloader::Associations` by their query and to run the query to load the records (which nicely avoids using `.first`).

This should not change any functionality but should be slightly faster due to being able to avoid calling `scope.to_sql` unless necessary.

cc @dinahshi @eileencodes @HParker

kamipo commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -4,16 +4,52 @@ module ActiveRecord
+          # This may cause extra collisions and thus `eql?` comparisons, but
+          # shouldn't be hugely significant.
+          def hash
+            [scope.table_name, scope.where_values_hash, association_key_name].hash
```
`scope.where_values_hash` is filtered to include only Equality and In conditions, not Between or NotEqual conditions etc, recommends `scope.values` or `scope.where_clause` as an alternative.

jhawthorn replied:
Unfortunately `WhereClause` and similar don't implement `hash` (they'll use `Object#hash` and basically hash by identity). This is why we use `scope.to_sql` above to compare, but I want to avoid needing to make that calculation in `hash`. This doesn't need to be exact (`eql?` does), just provide a good enough key that we don't have _too_ many collisions.

kamipo replied:
How about adding `WhereClause#hash`?

```diff
diff --git a/activerecord/lib/active_record/relation/where_clause.rb b/activerecord/lib/active_record/relation/where_clause.rb
index 37fbb12575..f75710adcc 100644
--- a/activerecord/lib/active_record/relation/where_clause.rb
+++ b/activerecord/lib/active_record/relation/where_clause.rb
@@ -72,6 +72,10 @@ def ast
         predicates.one? ? predicates.first : Arel::Nodes::And.new(predicates)
       end
 
+      def hash
+        [self.class, predicates].hash
+      end
+
       def ==(other)
         other.is_a?(WhereClause) &&
           predicates == other.predicates
```

jhawthorn replied:
Awesome! Didn't expect it to be that simple. I might be able to replace the rest of the `eql?` as well.

kamipo commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -4,16 +4,52 @@ module ActiveRecord
+            return false unless scope.preload_values == other.scope.preload_values
+            return false unless scope.includes_values == other.scope.includes_values
+            return true if scope.empty_scope? && other.scope.empty_scope?
+            return false unless scope.where_values_hash == other.scope.where_values_hash
```
We can use `scope.where_clause` here.

```suggestion
            return false unless scope.where_clause == other.scope.where_clause
```

jhawthorn committed:
Implement hash equality for WhereClause

Co-authored-by: Ryuta Kamizono <kamipo@gmail.com>

kamipo commented on activerecord/lib/active_record/associations/preloader/association.rb:
```diff
@@ -4,16 +4,43 @@ module ActiveRecord
+              scope.values == other.scope.values
+          end
+
+          # hash intentionally avoids scope.to_sql, so that it is only
+          # calculated when needed.
```
Is this comment still needed? `scope.to_sql` no longer exists in `eql?`.

kamipo commented on activerecord/lib/active_record/relation/where_clause.rb:
```diff
@@ -76,6 +76,11 @@ def ==(other)
         other.is_a?(WhereClause) &&
           predicates == other.predicates
       end
+      alias_method :eql?, :==
```
```suggestion
      alias :eql? :==
```

jhawthorn committed:
Add Preloader::Association::LoaderQuery

This class represents the query a Preloader::Association uses to load
its records. This replaces the previous concept of a grouping_key and
serves both to group Preloader::Associations by their query and to run
the query to load the records (which nicely avoids using .first).

This should not change any functionality but should be slightly faster
due to being able to avoid calling `scope.to_sql` unless necessary.
